### PR TITLE
DEP: Actually deprecate np.dtype('O8')

### PR DESCRIPTION
**Pull request by eric-wieser**
The code before tried to:
* Deprecate `np.dtype('O8')`
* Silently allow loading pickles containing `'O8'`

But actually due to a logic bug:
* Silently allowed `np.dtype('O8')`
* Deprecated loading pickles containing `'O8'`

Since we already deprecated the pickles for 8 years by accident, we may as well just deprecate both uses now.

### CODE REVIEW

File: numpy/core/src/multiarray/conversion_utils.c [hunk 1]
```diff
@@ -1064,14 +1052,11 @@ PyArray_TypestrConvert(int itemsize, int gentype)
              */
             if (itemsize == 4 || itemsize == 8) {
                 int ret = 0;
```
Comment (seberg) [inline numpy/core/src/multiarray/conversion_utils.c]:
```suggestion
```

**Review by seberg — DISMISSED** [review]
LGTM, will probably just commit the suggestion and merge myself later. Fun typo, just shows that even simple code is better with tests :).

**Comment by seberg** [comment]
Hmm, a bit more fallout than I thought, there is an actual tests unpickling an `O8` dtype that fails currently.

**Comment by eric-wieser** [comment]
> Hmm, a bit more fallout than I thought, there is an actual tests unpickling an `O8` dtype that fails currently.

I was under the impression that when unpickling the warning was already emitted. Am I mistaken?

**Comment by seberg** [comment]
I am as confused as you are, but round tripping tests for `npz` are failing for example:
https://dev.azure.com/numpy/numpy/_build/results?buildId=8798&view=logs&j=25bb66cf-4a16-533e-490c-8da4b5a3ec04&t=97446ee6-ebce-5662-24e5-f82eed312005&l=802

Probably this is not array but the single `dtype` pickle. However, the roundtripping should certainly work, so I guess the state setting may be off, since it should not be storing `O8` to begin with?

**Comment by eric-wieser** [comment]
Here's the issue:
```python
>>> np.dtype(object).__reduce__()
(numpy.dtype, ('O8', False, True), (3, '|', None, None, None, -1, -1, 63))
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 2]
```diff
@@ -930,18 +930,6 @@ PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals)
...
- * WARNING: This flag is a bad idea, but was the only way to both
- *   1) Support unpickling legacy pickles with object types.
- *   2) Deprecate (and later disable) usage of O4 and O8
- *
- * The key problem is that the pickled representation unpickles by
- * directly calling the dtype constructor, which has no way of knowing
- * that it is in an unpickle context instead of a normal context without
- * evil global state like we create here.
- */
-NPY_NO_EXPORT int evil_global_disable_warn_O4O8_flag = 0;
```
Comment (anirudh2290) [inline numpy/core/src/multiarray/conversion_utils.c:943]: can we also get rid of extern declaration in conversion_utils.h

**Comment by mattip** [comment]
There are still failing tests

**Comment by charris** [comment]
@eric-wieser Ping. Do you want to keep this open? You have a bunch of PRs, it would be helpful if you looked through them and decided which you would like to keep open.

**Pull request closed without merging**